File: server/src/signal.rs
use serde::{Deserialize, Serialize};

use crate::role::Role;

#[derive(Debug, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum SignalMessage
{
    #[serde(rename = "join")]
    Join { role: Role },

    #[serde(rename = "offer")]
    Offer { sdp: String, target_id: Option<String> },
    
    #[serde(rename = "answer")]
    Answer { sdp: String, target_id: Option<String> },

    #[serde(rename = "ice-candidate")]
    Ice { candidate: String, target_id: Option<String> },
}

File: client/src/signal.ts
export type SignalMessage =
  | { type: "join", role: "teacher" | "student" }
  | { type: "offer", sdp: string, target_id?: string | null }
  | { type: "answer", sdp: string, target_id?: string | null }
  | { type: "ice-candidate", candidate: string, target_id?: string | null }
  | { type: "peers-joined", peer_id: string }
